File: SIKEncrypt/src/main/cpp/utils/PaddingUtils.cpp
//
// Created by zhouw on 2023/6/26.
//

#include "PaddingUtils.h"

std::vector<uint8_t> PKCS5Padding(std::vector<uint8_t>& data, size_t block_size) {
    size_t length = block_size - (data.size() % block_size);
    data.insert(data.end(), length, (uint8_t)length);
    return data;
}

std::vector<uint8_t> PKCS5Unpadding(std::vector<uint8_t>& data) {
    if (data.empty()) {
        // Handle error: data is empty
        return data;
    }

    size_t length = data.back();
    if (length == 0 || length > data.size()) {
        // Handle error: invalid padding length
        return data;
    }

    // Check that all padding bytes have the correct value
    for (size_t i = 1; i <= length; ++i) {
        if (data[data.size() - i] != length) {
            // Handle error: invalid padding bytes
            return data;
        }
    }

    data.erase(data.end() - length, data.end());
    return data;
}

File: SIKEncrypt/src/main/java/com/sik/sikencrypt/PaddingUtils.kt
package com.sik.sikencrypt

object PaddingUtils {

    /**
     * Applies PKCS#5 padding to the given data.
     *
     * @param data       The data to be padded.
     * @param blockSize  The block size, in bytes.
     * @return A new byte array containing the padded data.
     */
    fun applyPKCS5Padding(data: ByteArray, blockSize: Int): ByteArray {
        require(blockSize > 0) { "Block size must be positive" }

        val paddingLength = blockSize - (data.size % blockSize)
        val paddedData = ByteArray(data.size + paddingLength)

        // Copy the original data
        System.arraycopy(data, 0, paddedData, 0, data.size)

        // Append the padding bytes
        for (i in 0 until paddingLength) {
            paddedData[data.size + i] = paddingLength.toByte()
        }

        return paddedData
    }

    /**
     * Removes PKCS#5 padding from the given data.
     *
     * @param data The padded data.
     * @return A new byte array containing the original data.
     * @throws IllegalArgumentException If the padding is invalid.
     */
    fun removePKCS5Padding(data: ByteArray): ByteArray {
        val paddingLength = data.last().toInt()
        if (paddingLength <= 0 || paddingLength > data.size) {
            throw IllegalArgumentException("Invalid PKCS#5 padding length")
        }

        // Check that all padding bytes have the correct value
        for (i in 1..paddingLength) {
            if (data[data.size - i] != paddingLength.toByte()) {
                throw IllegalArgumentException("Invalid PKCS#5 padding bytes")
            }
        }

        return data.copyOfRange(0, data.size - paddingLength)
    }
}
